paddle/fluid/operators/conv_op.h

```diff
@@ -393,6 +393,210 @@ class GemmConvGradKernel : public framework::OpKernel<T> {
+                      "It must use CPUPlace.");
+    const Tensor* X = ctx.Input<Tensor>("Input");
+    const Tensor* dY = ctx.Input<Tensor>("DOutput");
+    const Tensor* ddX = ctx.Input<Tensor>("DDInput");
+    const Tensor* ddW_in = ctx.Input<Tensor>("DDFilter");
+
+    Tensor* ddY = ctx.Output<Tensor>("DDOutput");
+    Tensor* dW = ctx.Output<Tensor>("DFilter");
+    Tensor* dX = ctx.Output<Tensor>("DInput");
+    Tensor W = *ctx.Input<Tensor>("Filter");
+
+    if (!ddY && !dW && !dX) return;
+    int groups = ctx.Attr<int>("groups");
+    std::vector<int> strides = ctx.Attr<std::vector<int>>("strides");
+    std::vector<int> paddings = ctx.Attr<std::vector<int>>("paddings");
+    std::vector<int> dilations = ctx.Attr<std::vector<int>>("dilations");
+
+    const int batch_size = static_cast<int>(X->dims()[0]);
+    std::vector<int64_t> filter_shape_vec(framework::vectorize(W.dims()));
+    std::vector<int64_t> output_shape_vec(framework::vectorize(dY->dims()));
+
+    size_t data_dim = filter_shape_vec.size() - 2;
+    std::vector<int64_t> col_shape_vec(1 + 2 * data_dim);
+    // col_shape [in_channel/group, kh, kw, oh, ow]
+    col_shape_vec[0] = X->dims()[1] / groups;
+    for (size_t j = 0; j < data_dim; ++j) {
+      col_shape_vec[j + 1] = filter_shape_vec[j + 2];
+      col_shape_vec[j + data_dim + 1] = output_shape_vec[j + 2];
+    }
+    framework::DDim col_shape(framework::make_ddim(col_shape_vec));
+    // col_matrix_shape [in_channel/group * kh * kw, oh * ow]
+    framework::DDim col_matrix_shape =
+        framework::flatten_to_2d(col_shape, data_dim + 1);
+    // input_shape [Cin, H, W]
+    framework::DDim input_shape =
+        framework::slice_ddim(X->dims(), 1, X->dims().size());
+    // filter_matrix_shape [Cout, Cin * kh * kw]
+    framework::DDim filter_matrix_shape = {W.dims()[0],
+                                           W.numel() / W.dims()[0]};
+
+    W.Resize(filter_matrix_shape);
+    framework::DDim output_matrix_shape = {
+        dY->dims()[1], dY->numel() / (dY->dims()[0] * dY->dims()[1])};
+    int in_step = static_cast<int>(X->dims()[1]) / groups;
+    int out_step = static_cast<int>(dY->dims()[1]) / groups;
+
+    bool is_expand = IsExpand(filter_shape_vec, strides, paddings, dilations);
+    Tensor col;
+    Tensor col_matrix;
+    if (is_expand) {
+      col = ctx.AllocateTmpTensor<T, DeviceContext>(col_shape, dev_ctx);
+      col_matrix.ShareDataWith(col);
+      col_matrix.Resize(col_matrix_shape);
+    }
+
+    math::SetConstant<DeviceContext, T> set_zero;
+    auto blas = math::GetBlas<DeviceContext, T>(dev_ctx);
+
+    // dx convolution double grad:  gemm + col2im(col2vol)
+    // dx = ddw * dy  ==> dx(N, Cin, H, W), ddw(Cout, Cin, kh, kw), dy(N, Cout,
+    // oH, oW)
+    if (dX && ddW_in) {
+      Tensor ddW =
+          ctx.AllocateTmpTensor<T, DeviceContext>(ddW_in->dims(), dev_ctx);
+      ddW.ShareDataWith(*ddW_in).Resize(filter_matrix_shape);
```

Comment:
Since use ShareDataWith,  above `ctx.AllocateTmpTensor` is unnecessary. Ony need to define a Tensor：

```
Tensor ddW;
```